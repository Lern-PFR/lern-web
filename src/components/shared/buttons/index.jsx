import PropTypes from 'prop-types';
import {
	primary,
	standard,
	subtle,
	dropdown,
	danger,
	iconButton,
	floatingActionButton,
	primaryLink,
	outlined,
	subtleLink,
} from 'theme/buttonStyles';
import DynamicButtonComponent from './DynamicButtonComponent';
import DynamicLinkButtonComponent from './DynamicLinkButtonComponent';

// @TODO: write link and subtleLink buttons once navigation has been set up.

/**
 * @name PrimaryButton
 * @description Button component to use as Call to action. Should only be used once per page (header not included).
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const PrimaryButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...primary} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name PrimaryLinkButton
 * @description Link button component to use as Call to action. Should only be used once per page (header not included).
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const PrimaryLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={primaryLink} {...primary} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name OutlinedButton
 * @description Button component to use as primary action when a CTA is already defined.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const OutlinedButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent linkStyle={primaryLink} {...outlined} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name OutlinedLinkButton
 * @description Link button component to use as primary when a CTA is already defined.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children The children to display inside this button.
 */
const OutlinedLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={primaryLink} {...outlined} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name StandardButton
 * @description Standard button component to be used for most use-cases.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const StandardButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...standard} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name SubtleButton
 * @description Button component to be used along a primary button for less crucial actions such as "cancel".
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const SubtleButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...subtle} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name SubtleLinkButton
 * @description Link button component to be used along a primary button for less used redirections.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const SubtleLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={subtleLink} {...subtle} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name DropdownButton
 * @description Button component to be used for dropdown menu trigger.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const DropdownButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...dropdown} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name DangerButton
 * @description Button component to use for final confirmation to a destructive action such as deleting data.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const DangerButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...danger} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name IconButton
 * @description Button component intended to display only the icon it contains, but behave as a regular button.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const IconButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...iconButton} {...otherProps} isIconOnly>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name FloatingActionButton
 * @description Button component intended to display a floating action button.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const FloatingActionButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...floatingActionButton} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

PrimaryButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

PrimaryLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

OutlinedButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

OutlinedLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

StandardButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

SubtleButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

SubtleLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

DropdownButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

DangerButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

IconButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

FloatingActionButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

export {
	FloatingActionButton,
	OutlinedButton,
	OutlinedLinkButton,
	PrimaryButton,
	PrimaryLinkButton,
	StandardButton,
	SubtleButton,
	SubtleLinkButton,
	DropdownButton,
	DangerButton,
	IconButton,
};
