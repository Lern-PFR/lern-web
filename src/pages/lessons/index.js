import LessonCreationPage from './LessonCreationPage';
import LessonEditionPage from './LessonEditionPage';

export {
	LessonCreationPage,
	LessonEditionPage,
};
