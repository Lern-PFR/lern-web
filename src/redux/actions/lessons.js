import * as LessonsApi from 'api/lessonsApi';
import i18next from 'i18next';
import { toast } from 'react-toastify';
import { fetchConcept } from './concepts';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the lessons redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_LESSON_REQUEST: '@LESSONS/FETCH_REQUEST',
	FETCH_LESSON_SUCCESS: '@LESSONS/FETCH_SUCCESS',
	FETCH_LESSON_FAILURE: '@LESSONS/FETCH_FAILURE',

	CREATE_LESSON_REQUEST: '@LESSONS/CREATE_REQUEST',
	CREATE_LESSON_SUCCESS: '@LESSONS/CREATE_SUCCESS',
	CREATE_LESSON_FAILURE: '@LESSONS/CREATE_FAILURE',

	UPDATE_LESSON_REQUEST: '@LESSONS/UPDATE_REQUEST',
	UPDATE_LESSON_SUCCESS: '@LESSONS/UPDATE_SUCCESS',
	UPDATE_LESSON_FAILURE: '@LESSONS/UPDATE_FAILURE',

	DELETE_LESSON_REQUEST: '@LESSONS/DELETE_REQUEST',
	DELETE_LESSON_SUCCESS: '@LESSONS/DELETE_SUCCESS',
	DELETE_LESSON_FAILURE: '@LESSONS/DELETE_FAILURE',

	CLEAR_LESSON_LIST: '@LESSONS/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific lesson fetching actions ////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchLessonRequest
 * @description Action triggered anytime a lesson fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchLessonRequest = () => ({ type: ActionTypes.FETCH_LESSON_REQUEST });

/**
  * @function
  * @name fetchLessonSuccess
  * @description Action triggered as a result to a successful lesson fetching API call.
  *
  * @author Timothée Simon-Franza
  *
  * @param {object} lesson : The lesson retrieved from the API.
  *
  * @returns {object}
  */
const fetchLessonSuccess = ({ lesson }) => ({
	type: ActionTypes.FETCH_LESSON_SUCCESS,
	payload: { lesson },
});

/**
 * @function
 * @name fetchLessonFailure
 * @description Action triggered as a result to a failed lesson fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchLessonFailure = (error) => ({
	type: ActionTypes.FETCH_LESSON_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Lesson creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createLessonRequest
 * @description Action triggered anytime a lesson creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createLessonRequest = () => ({ type: ActionTypes.CREATE_LESSON_REQUEST });

/**
 * @function
 * @name createLessonSuccess
 * @description Action triggered as a result to a successfult lesson creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} lesson : The newly created lesson.
 *
 * @returns {object}
 */
const createLessonSuccess = ({ lesson }) => ({
	type: ActionTypes.CREATE_LESSON_SUCCESS,
	payload: { lesson },
});

/**
 * @function
 * @name createLessonFailure
 * @description Action triggered as a result to a failed lesson creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const createLessonFailure = (error) => ({
	type: ActionTypes.CREATE_LESSON_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Lesson edition actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateLessonRequest
 * @description Action triggered anytime a lesson edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateLessonRequest = () => ({ type: ActionTypes.UPDATE_LESSON_REQUEST });

/**
 * @function
 * @name updateLessonSuccess
 * @description Action triggered as a result to a successful lesson edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} lesson : The updated lesson object.
 *
 * @returns {object}
 */
const updateLessonSuccess = ({ lesson }) => ({
	type: ActionTypes.UPDATE_LESSON_SUCCESS,
	payload: { lesson },
});

/**
 * @function
 * @name updateLessonFailure
 * @description Action triggered as a result to a failed lesson edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const updateLessonFailure = (error) => ({
	type: ActionTypes.UPDATE_LESSON_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Lesson deletion actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteLessonRequest
 * @description Action triggered anytime a lesson deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteLessonRequest = () => ({ type: ActionTypes.DELETE_LESSON_REQUEST });

/**
 * @function
 * @name deleteLessonSuccess
 * @description Action triggered as a result to a successful lesson deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} lesson : The deleted lesson object.
 *
 * @returns {object}
 */
const deleteLessonSuccess = ({ lesson }) => ({
	type: ActionTypes.DELETE_LESSON_SUCCESS,
	payload: { lesson },
});

/**
 * @function
 * @name deleteLessonFailure
 * @description Action triggered as a result to a failed lesson deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteLessonFailure = (error) => ({
	type: ActionTypes.DELETE_LESSON_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Lesson list clearing actions ///////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name clearLessonListAction
 * @description Action triggered anytime the lesson list is not needed anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const clearLessonListAction = () => ({ type: ActionTypes.CLEAR_LESSON_LIST });

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchLesson
 * @description Method used to fetch the latest version of a specified lesson.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} lessonId : The id of the lesson we want to retrieve.
 */
export const fetchLesson = (lessonId) => (dispatch) => {
	dispatch(fetchLessonRequest());

	return LessonsApi.fetchLessonById(lessonId)
		.then(({ lesson }) => dispatch(fetchLessonSuccess({ lesson })))
		.catch((error) => dispatch(fetchLessonFailure(error)));
};

/**
 * @function
 * @name createLesson
 * @description Method used to create a new lesson in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} lessonData : The data to create the new lesson from.
 */
export const createLesson = (lessonData) => (dispatch) => {
	dispatch(createLessonRequest());

	return LessonsApi.createLesson(lessonData)
		.then(({ lesson }) => dispatch(createLessonSuccess({ lesson })))
		.catch((error) => dispatch(createLessonFailure(error)));
};

/**
 * @function
 * @name updateLesson
 * @description Method used to update an existing lesson instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} lessonData	: The data to update the lesson with.
 * @param {string} lessonId		: The id to identify the lesson to update.
 */
export const updateLesson = (lessonData, lessonId) => (dispatch) => {
	dispatch(updateLessonRequest());

	return LessonsApi.updateLesson(lessonData, lessonId)
		.then(({ lesson }) => dispatch(updateLessonSuccess({ lesson })))
		.catch((error) => dispatch(updateLessonFailure(error)));
};

/**
 * @function
 * @name deleteLesson
 * @description Method used to remove an existing lesson from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} lessonId : The id identifying the lesson to remove.
 */
export const deleteLesson = (lessonId) => (dispatch) => {
	dispatch(deleteLessonRequest());

	return LessonsApi.deleteLesson(lessonId)
		.then((lesson) => {
			dispatch(deleteLessonSuccess({ lesson }));
			toast.success(i18next.t('lessons.deletion.toasts.success', { name: lesson.title }));
			dispatch(fetchConcept(lesson.conceptId));
		})
		.catch((error) => dispatch(deleteLessonFailure(error)));
};

/**
 * @function
 * @name clearLessonList
 * @description Method used to clear the lesson list stored in memory when the application does not need it anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Promise.resolve}
 */
export const clearLessonList = () => (dispatch) => {
	dispatch(clearLessonListAction());

	return Promise.resolve();
};
