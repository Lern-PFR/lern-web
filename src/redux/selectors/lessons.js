import { sortBy } from 'lodash';
import { createSelector } from 'reselect';
import { getUserAnswer } from './userAnswers';

// const extractFirstQuestionFromLesson = (lesson) => (lesson?.exercises?.[0]?.questions?.[0]);

/**
 * @function
 * @name extractFirstQuestionFromLesson
 * @description Extracts the first question of the given lesson if it exists.
 *
 * @param {object} lesson The lesson to extract a question from.
 *
 * @return {object|undefined}
 */
const extractFirstQuestionFromLesson = createSelector(
	getUserAnswer,
	(_, lesson) => lesson ?? undefined,
	(userAnswer, lesson) => {
		const question = lesson?.exercises?.[0]?.questions?.[0];

		if (!question) {
			return undefined;
		}

		if (question && userAnswer && question.id === userAnswer.questionId && question.answers) {
			question.answers = question.answers.map((answer) => ({
				...answer,
				isUserAnswer: answer.id === userAnswer.answerId,
			}));
		}

		return {
			...question,
			answers: sortBy(question.answers ?? [], 'text'),
		};
	}
);

export {
	// eslint-disable-next-line import/prefer-default-export
	extractFirstQuestionFromLesson,
};
