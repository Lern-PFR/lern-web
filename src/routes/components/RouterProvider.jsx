import PropTypes from 'prop-types';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

/**
 * @name history
 * @description The global history object, handling the navigation for the whole application
 *
 * @author Yann Hodiesne
 */
export const history = createBrowserHistory();

/**
 * @name RouterProvider
 * @description A component providing a Router from react-router, with the ability to export its history and block navigation
 *
 * @author Yann Hodiesne
 *
 * @param {*} children : The children to render inside the Router
 */
const RouterProvider = ({ children }) => (
	<Router history={history}>
		{children}
	</Router>
);

RouterProvider.propTypes = {
	children: PropTypes.node.isRequired,
};

export default RouterProvider;
