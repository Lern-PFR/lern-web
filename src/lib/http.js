import localstorage from './localstorage';
import history from './history'
import { logoutAction } from '../redux/actions/auth';
import store from '../redux/store/configureStore';

const { dispatch, getState } = store;

/**
 * Generates HTTP headers.
 * 
 * If the user is authenticated, retrieves its jwt authentication token and add it to the header list.
 */
const getHeaders = () => {
    const headers = new Headers();
    const user = JSON.parse(localstorage.getItem('user'));
    
    headers.append('Content-Type', 'application/json');
  
    if (user && user.token) {
        headers.append('Authorization', `Bearer ${user.token}`);
    }
  
    return headers;
};

/**
 * Returns a query string from the object in parameter.
 * 
 * @param {object} params : The object to create the query string from.
 */
const objectToQS = (params) => Object
    .entries(params)
    .map(([key, value]) => {
        if (value !== undefined && value !== null) {
            return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
        }

        return `${encodeURIComponent(key)}`;
    })
    .join('&');

/**
 * Generates a POST HTTP query from the url and params object in parameters.
 * 
 * @param {string} url      : The url to target.
 * @param {object} params   : The parameters to inject into the query.
 */
export const post = (url, params) => fetch(url, {
    method: 'POST',
    body: JSON.stringify(params),
    headers: getHeaders(),
});

/**
 * Generates a GET HTTP query from the url and params object in parameters.
 * 
 * If the params object is set, converts its content into a query string.
 * 
 * @param {string} url      : The url to target.
 * @param {object} params   : The parameters to inject into the query.
 */
export const get = (url, params) => {
    const qs = params ? `?${objectToQS(params)}` : '';

    return fetch(`${url}${qs}`, {
        headers: getHeaders(),
    });
};

/**
 * Generates a PUT HTTP query from the url and params object in parameters.
 * 
 * @param {string} url       : The url to target.
 * @param {object} params    : The parameters to inject into the query.
 */
export const put = (url, params) => fetch(url, {
    method: 'PUT',
    body: JSON.stringify(params),
    headers: getHeaders(),
});

/**
 * Generates a DELETE HTTP query from the url in parameters.
 * 
 * @param {string} url 
 */
export const del = (url) => fetch(url, {
    method: 'PUT',
    headers: getHeaders(),
});

/**
 * Handles the response from an HTTP query.
 * 
 * If a trigger is passed to the method, dispatch it using redux.
 * 
 * @param {*} trigger : The redux action to dispatch.
 */
export const handleResponse = (trigger) => (response) => {
    if (response.status === 204) {
      if (trigger) {
        dispatch(trigger());
      }
  
      return Promise.resolve();
    }
  
    if (response.status === 403 || response.status === 413 || response.status >= 500) {
      return Promise.reject(response);
    }
  
    return response.json().then((res) => {
        if (response.status === 401 && res.message !== 'Bad credentials') {
            dispatch(logoutAction());
            history.push('/');
    
            return Promise.reject({
                ...res,
                status: response.status,
            });
        }
    
        if (response.status >= 300) {
            return Promise.reject({
                ...res,
                status: response.status,
            });
        }
    
        if (trigger) {
            dispatch(trigger(res));
        }
    
        return Promise.resolve(res);
    });
};