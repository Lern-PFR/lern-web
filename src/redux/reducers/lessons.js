import { ActionTypes } from 'redux/actions/lessons';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_LESSON_REQUEST:
		case ActionTypes.CREATE_LESSON_REQUEST:
		case ActionTypes.UPDATE_LESSON_REQUEST:
		case ActionTypes.DELETE_LESSON_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_LESSON_FAILURE:
		case ActionTypes.CREATE_LESSON_FAILURE:
		case ActionTypes.UPDATE_LESSON_FAILURE:
		case ActionTypes.DELETE_LESSON_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_LESSON_SUCCESS:
		case ActionTypes.DELETE_LESSON_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.UPDATE_LESSON_SUCCESS:
		case ActionTypes.FETCH_LESSON_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.lesson],
				totalCount: 1,
			};
		case ActionTypes.CLEAR_LESSON_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
