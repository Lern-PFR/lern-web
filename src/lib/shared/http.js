import { history } from 'routes/components/RouterProvider';
import session from 'lib/shared/session';
import routes from 'routes';

/**
 * @constant
 * @name baseUrl
 * @description The base url to append to in order to target and API endpoint.
 * @type {string}
 *
 * @author Timothée Simon-Franza
 *
 * Note : Its value is set using environment variables set defined in this .env file and forwarded by docker-compose.
 */
export const baseUrl = process.env.REACT_APP_API_URL;

/**
 * @function
 * @name getHeaders
 * @description Generates and returns a header bundle from the current session's data.
 *
 * @author Timothée Simon-Franza
 * @author Yann Hodiesne
 *
 * @returns {Header} the generated header bundle.
 */
export const getHeaders = () => {
	const headers = new Headers();

	headers.append('Content-Type', 'application/json');

	if (session.exists()) {
		headers.append('Authorization', `Bearer ${session.get()}`);
	}

	return headers;
};

/**
 * @function
 * @name objectToQS
 * @description Converts the object in parameter to a query string.
 *
 * @author Timothée Simon-Franza
 * @author Yann Hodiesne
 *
 * @param {object} params : The object to convert to QS.
 *
 * @returns {string}
 */
export const objectToQS = (params) => Object
	.entries(params)
	.map(([key, value]) => {
		if (value !== undefined && value !== null) {
			return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
		}

		return `${encodeURIComponent(key)}`;
	})
	.join('&');

/**
 * @function
 * @name handleResponse
 * @description Automatically deserializes and handles the received response object.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} response : The response to handle.
 *
 * @returns {Promise | object}
 */
export const handleResponse = (response) => {
	if (response.status === 204) {
		return Promise.resolve();
	}

	if (response.status === 401 && session.exists()) {
		history.push(routes.auth.logout);
	}

	if (response.status === 401 || response.status === 403 || response.status === 409 || response.status === 413 || response.status >= 500) {
		return Promise.reject(response);
	}

	return response.json();
};

/**
 * @function
 * @name handleError
 * @description Automatically handles the received error to return a formatted object.
 *
 * @author Yann Hodiesne
 *
 * @param {object} error : The error to handle.
 *
 * @returns {Promise.reject}
 */
export const handleError = (error) => {
	const {
		status,
		statusText: message,
	} = error;

	return Promise.reject({ status, message });
};

/**
 * @function
 * @name post
 * @description Creates a POST method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url	: The url to target.
 * @param {object} body	: The body to include.
 *
 * @returns {Promise | object}
 */
export const post = (url, body) => fetch(`${baseUrl}${url}`, {
	method: 'POST',
	body: JSON.stringify(body),
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);

/**
 * @function
 * @name get
 * @description Creates a GET method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url		: The url to target.
 * @param {objcet} qsObject	: The object to convert into a query string.
 *
 * @returns {Promise | object}
 */
export const get = (url, qsObject) => {
	const queryString = (qsObject && Object.keys(qsObject).length !== 0) ? `?${objectToQS(qsObject)}` : '';

	return fetch(`${baseUrl}${url}${queryString}`, {
		method: 'GET',
		headers: getHeaders(),
	}).then(handleResponse).catch(handleError);
};

/**
 * @function
 * @name put
 * @description Creates a PUT method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url : The url to target.
 * @param {object} body : The body to include.
 */
export const put = (url, body) => fetch(`${baseUrl}${url}`, {
	method: 'PUT',
	body: JSON.stringify(body),
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);

/**
 * @function
 * @name del
 * @description Creates a DELETE method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url : The url to target.
 */
export const del = (url) => fetch(`${baseUrl}${url}`, {
	method: 'DELETE',
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);
