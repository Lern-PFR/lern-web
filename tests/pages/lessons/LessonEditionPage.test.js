import { mount, shallow } from 'enzyme';
import { LessonEditionPage } from 'pages/lessons';

import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import routeData from 'react-router';
import { baseUrl } from 'lib/shared/http';

import { ActionTypes as LessonsActionTypes } from 'redux/actions/lessons';
import RouterProvider from 'routes/components/RouterProvider';
import fetchMock from 'fetch-mock';

const mockStore = configureMockStore([thunk]);

describe('Lesson edition page', () => {
	let store;

	beforeEach(() => {
		jest.spyOn(routeData, 'useParams').mockReturnValue({ lessonId: 'dummy_lesson_id' });
	});

	afterEach(() => {
		jest.restoreAllMocks();
		jest.resetAllMocks();
	});

	describe('Snapshot testing', () => {
		it('should match previous snapshot without the required lesson in state', () => {
			store = mockStore({});

			const sut = (
				<Provider store={store}>
					<RouterProvider>
						<LessonEditionPage />
					</RouterProvider>
				</Provider>
			);

			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});

		it('should match previous snapshot with the required lesson, module and subject in state', () => {
			store = mockStore({
				subjects: {
					all: [{
						id: 'dummy_subject_id',
						title: 'dummy_subject_title',
						description: 'dummy_subject_desc',
						modules: [{
							id: 'dummy_module_id',
							title: 'dummy_module_title',
							description: 'dummy_module_desc',
							subjectId: 'dummy_subject_id',
							order: 0,
							concepts: [{
								id: 'dummy_concept_id',
								title: 'dummy_concept_title',
								description: 'dummy_concept_desc',
								moduleId: 'dummy_module_id',
								order: 0,
							}],
						}],
					}],
				},
				modules: [{
					id: 'dummy_module_id',
					title: 'dummy_module_title',
					description: 'dummy_module_desc',
					subjectId: 'dummy_subject_id',
					order: 0,
					concepts: [{
						id: 'dummy_concept_id',
						title: 'dummy_concept_title',
						description: 'dummy_concept_desc',
						moduleId: 'dummy_module_id',
						order: 0,
					}],
				}],
				concepts: [{
					id: 'dummy_concept_id',
					title: 'dummy_concept_title',
					description: 'dummy_concept_desc',
					moduleId: 'dummy_module_id',
					order: 0,
				}],
				lessons: [{
					id: 'dummy_lesson_id',
					title: 'dummy_lesson_title',
					description: 'dummy_lesson_desc',
					content: 'dummy_lesson_content',
					conceptId: 'dummy_concept_id',
					order: 0,
				}],
			});

			const sut = (
				<Provider store={store}>
					<RouterProvider>
						<LessonEditionPage />
					</RouterProvider>
				</Provider>
			);

			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('dispatched method calls', () => {
		it('should call the fetchLesson action creator on mount', async () => {
			const subjectHttpResponse = {
				status: 200,
				body: { id: 'dummy_subject_id', name: 'dummy_subject_name' },
				headers: { 'content-type': 'application/json' },
			};

			const moduleHttpResponse = {
				status: 200,
				body: { id: 'dummy_module_id', name: 'dummy_module_name', subjectId: 'dummy_subject_id' },
				headers: { 'content-type': 'application/json' },
			};

			const conceptHttpResponse = {
				status: 200,
				body: { id: 'dummy_concept_id', name: 'dummy_concept_name', moduleId: 'dummy_module_id' },
				headers: { 'content-type': 'application/json' },
			};

			const lessonHttpResponse = {
				status: 200,
				body: {
					id: 'dummy_lesson_id',
					title: 'dummy_lesson_title',
					description: 'dummy_lesson_desc',
					content: 'dummy_lesson_content',
					conceptId: 'dummy_concept_id',
					order: 0,
				},
				headers: { 'content-type': 'application/json' },
			};

			fetchMock.get(`${baseUrl}/api/lessons/dummy_lesson_id`, lessonHttpResponse);
			fetchMock.get(`${baseUrl}/api/concepts/dummy_concept_id`, conceptHttpResponse);
			fetchMock.get(`${baseUrl}/api/modules/dummy_module_id`, moduleHttpResponse);
			fetchMock.get(`${baseUrl}/api/subjects/dummy_subject_id`, subjectHttpResponse);

			mount(
				<Provider store={store}>
					<RouterProvider>
						<LessonEditionPage />
					</RouterProvider>
				</Provider>
			);

			const expectedActions = [
				{ type: LessonsActionTypes.FETCH_LESSON_REQUEST },
			];

			expect(store.getActions()).toEqual(expectedActions);
		});
	});
});
