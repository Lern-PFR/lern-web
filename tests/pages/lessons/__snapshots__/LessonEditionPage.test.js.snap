// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson edition page Snapshot testing should match previous snapshot with the required lesson, module and subject in state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RouterProvider>
    <LessonEditionPage />
  </RouterProvider>
</ContextProvider>
`;

exports[`Lesson edition page Snapshot testing should match previous snapshot without the required lesson in state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RouterProvider>
    <LessonEditionPage />
  </RouterProvider>
</ContextProvider>
`;
